pkg/services/ngalert/api/fork_ruler.go

```diff
@@ -2,26 +2,57 @@ package api
 
 import (
 	"fmt"
+	"time"
 
 	"github.com/grafana/grafana/pkg/api/response"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/datasources"
 	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	GrafanaBackend = "grafana"
+	ProxyBackend   = "proxy" // metric cardinality is too high to enumerate all non-grafana backends
 )
 
 // ForkedRuler will validate and proxy requests to the correct backend type depending on the datasource.
 type ForkedRuler struct {
 	LotexRuler, GrafanaRuler RulerApiService
 	DatasourceCache          datasources.CacheService
+
+	// metrics
+	duration *prometheus.HistogramVec
 }
 
 // NewForkedRuler implements a set of routes that proxy to various Cortex Ruler-compatible backends.
-func NewForkedRuler(datasourceCache datasources.CacheService, lotex, grafana RulerApiService) *ForkedRuler {
-	return &ForkedRuler{
+func NewForkedRuler(datasourceCache datasources.CacheService, lotex, grafana RulerApiService, reg prometheus.Registerer) *ForkedRuler {
+	r := &ForkedRuler{
 		LotexRuler:      lotex,
 		GrafanaRuler:    grafana,
 		DatasourceCache: datasourceCache,
+		duration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "grafana",
+			Name:      "alerting_ruler_api_duration_seconds",
+			Help:      "Histogram of latencies affecting the Unified Alerting Ruler API",
+			Buckets:   prometheus.DefBuckets,
+		}, []string{"backend", "status"}),
```

Comment:
I feel we should include route, w/o route it becomes pretty hard to answer the important questions (e.g. which route is the slowest?) - I understand the concern of cardinality in a multi-tenant environment but I fear that's a problem that Grafana shouldn't be attempting to solve... I'm willing to be swayed the other way for the purpose of practicality though.

Reply:
I suspect you might be right here, although I'm always wary of cardinality :boom: . I've been toying with the idea of adding more granular events to logs, but I think I'm optimizing early. Let's go per-route initially as it's what we _want_ and then adjust accordingly if we see problems.